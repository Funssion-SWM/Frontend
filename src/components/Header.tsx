'use client';

import exampleImg from '../../public/img/profile.png';
import Link from 'next/link';
import { checkUser, logout } from '@/service/auth';
import Image from 'next/image';
import { useEffect, useRef, useState } from 'react';
import { useDetectOutsideClick } from '@/hooks/useDeleteOutsideClick';
import { useRouter } from 'next/navigation';
import BlueBtn from './BlueBtn';

export default function Header() {
  const router = useRouter();
  const dropdownRef = useRef<HTMLElement>(null);
  const [isActive, setIsActive] = useDetectOutsideClick(dropdownRef, false);
  const [isLogin, setIsLogin] = useState(null);

  useEffect(() => {
    async function first() {
      1;
      await checkUser()
        .then((res) => {
          // if (!res.ok) throw new Error('error!!');
          console.log(res);
          return res.json();
        })
        .then((data) => {
          console.log(data);
          setIsLogin(data.isLogin);
        })
        .catch((err) => console.error(err));
    }
    first();
  }, [isActive]);

  return (
    <header className=" border-b-2">
      <div className="flex justify-between items-center py-4 px-10 max-w-screen-xl m-auto">
        <h1
          className="text-2xl font-bold cursor-pointer"
          onClick={() => {
            router.push('/');
            router.refresh();
          }}
        >
          Inforum
        </h1>

        {/* <nav className="flex gap-4">
        <Link href="/memos">Memos</Link>
        <Link href="/stories">Stories</Link>
      </nav> */}
        {isLogin === true && (
          <nav className="flex items-center gap-3 relative" ref={dropdownRef}>
            <button onClick={() => setIsActive((pre) => !pre)}>
              <Image
                src={exampleImg}
                alt="exampleImg"
                height={30}
                className="rounded-full"
              />
            </button>
            <nav
              className={`absolute top-10 bg-white flex flex-col gap-1 rounded-lg shadow-inner ${
                isActive ? 'visible' : 'invisible'
              }`}
            >
              <button
                className="hover:bg-gray-200 p-2 rounded-t-lg"
                onClick={() => {
                  checkUser()
                    .then((res) => {
                      if (!res.ok) throw new Error('error!!');
                      return res.json();
                    })
                    .then((data) => {
                      router.push(`/me/${data.id}`);
                    })
                    .catch(console.error);
                  setIsActive(false);
                }}
              >
                프로필
              </button>
              <button
                className="hover:bg-gray-200 p-2 rounded-b-lg"
                onClick={() => {
                  logout()
                    .then((res) => {
                      console.log(res);
                      if (!res.ok) throw new Error('error!!');
                      setIsActive(false);
                      router.push('/');
                      router.refresh();
                    })
                    .catch(console.error);
                }}
              >
                로그아웃
              </button>
            </nav>
            <BlueBtn
              text={'+글쓰기'}
              onClick={() => {
                router.push('/create/memo');
                setIsActive(false);
              }}
            />
          </nav>
        )}
        {isLogin === false && (
          <Link href="/login">
            <button className=" bg-blue-500 text-white px-2 py-1 rounded-2xl">
              회원가입/로그인
            </button>
          </Link>
        )}
      </div>
    </header>
  );
}
