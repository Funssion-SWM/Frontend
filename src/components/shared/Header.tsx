'use client';

import basicProfileImg from '../../assets/profile.svg';
import Link from 'next/link';
import { checkUser, logout } from '@/service/auth';
import searchIcon from '@/assets/icons/icon_search_32.svg';
import Image from 'next/image';
import { useContext, useRef } from 'react';
import { useDetectOutsideClick } from '@/hooks/useDeleteOutsideClick';
import { useRouter } from 'next/navigation';
import BlueBtn from './btn/BlueBtn';
import { ModalContext } from '@/context/ModalProvider';
import logo from '@/assets/inforum_logo.png';
import { useScrollDirection } from '@/hooks/useScrollDirection';

type Props = {
  isLogin: boolean;
  profileImageFilePath: string | undefined;
};

export default function Header({ isLogin, profileImageFilePath }: Props) {
  const router = useRouter();
  const dropdownRef = useRef<HTMLElement>(null);
  const [isActive, setIsActive] = useDetectOutsideClick(dropdownRef, false);
  const { open } = useContext(ModalContext);
  const scrollDirection = useScrollDirection();

  return (
    <header
      className={`border-b-[1px] border-soma-grey-40 sticky top-0 bg-white z-10 transition-all ease-in-out duration-300 ${
        scrollDirection === 'down' && 'opacity-0 invisible'
      }`}
    >
      <div className="flex justify-between items-center py-4 px-5 max-w-screen-xl m-auto h-[70px]">
        <Image
          src={logo}
          alt="logo"
          width={120}
          onClick={() => {
            router.push('/memos');
            router.refresh();
          }}
          className="cursor-pointer"
        />
        {isLogin ? (
          <nav className="flex items-center gap-3 relative" ref={dropdownRef}>
            <button onClick={() => router.push('/search/form')}>
              <Image
                className="cursor-pointer"
                src={searchIcon}
                alt="search_icon"
              />
            </button>

            <button onClick={() => setIsActive((pre) => !pre)}>
              <Image
                src={profileImageFilePath ?? basicProfileImg}
                alt="profileImg"
                width={32}
                height={32}
                className="rounded-full w-8 h-8 object-cover"
              />
            </button>

            <nav
              className={`absolute top-10 bg-white flex flex-col gap-1 rounded-lg shadow-inner z-10 ${
                isActive ? 'visible' : 'invisible'
              }`}
            >
              <button
                className="hover:bg-gray-200 p-2 rounded-t-lg"
                onClick={() => {
                  checkUser().then((data) => {
                    router.push(`/me/${data.id}`);
                    router.refresh();
                  });
                  setIsActive(false);
                }}
              >
                프로필
              </button>
              <button
                className="hover:bg-gray-200 p-2 rounded-t-lg"
                onClick={() => {
                  checkUser().then((data) =>
                    router.push(`/me/${data.id}/drafts`)
                  );
                  setIsActive(false);
                }}
              >
                임시 글
              </button>
              <button
                className="hover:bg-gray-200 p-2 rounded-b-lg"
                onClick={() => {
                  setIsActive(false);
                  open('로그아웃 하시겠습니까?', () => {
                    logout().then(() => {
                      router.push('/memos');
                      router.refresh();
                    });
                  });
                }}
              >
                로그아웃
              </button>
            </nav>
            <BlueBtn
              text={'글쓰기'}
              onClick={() => {
                router.push('/create/memo');
                setIsActive(false);
              }}
            />
          </nav>
        ) : (
          <nav className="flex items-center gap-3">
            <button onClick={() => router.push('/search/form')}>
              <Image
                className="cursor-pointer"
                src={searchIcon}
                alt="search_icon"
              />
            </button>
            <Link href="/login">
              <BlueBtn text="로그인" onClick={() => {}} />
            </Link>
          </nav>
        )}
      </div>
    </header>
  );
}
