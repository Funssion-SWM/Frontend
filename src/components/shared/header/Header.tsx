import { checkUser, getUserInfo } from '@/service/auth';
import LogoButton from './LogoButton';
import Navigater from './Navigater';
import { cookies } from 'next/headers';
import { ACCESS_TOKEN, REFRESH_TOKEN } from '@/constants/general';
import { getNotificationsTop30 } from '@/service/notification';
import Submenu1 from './SubMenu1';
import Submenu2 from './SubMenu2';
import ScrollAnimationCover from './ScrollAnimationCover';
import NavigatorInMobile from './NavigatorInMobile';

type Props = {
  currentPage?: 'memos' | 'questions' | 'series';
};

export default async function Header({ currentPage }: Props) {
  const accessToken = cookies().get(ACCESS_TOKEN)?.value;
  const refreshToken = cookies().get(REFRESH_TOKEN)?.value;
  const cookie = `${ACCESS_TOKEN}=${accessToken}; ${REFRESH_TOKEN}=${refreshToken}`;

  const { id, isLogin, authority } = await checkUser(cookie);

  const { profileImageFilePath } = isLogin
    ? await getUserInfo(id)
    : { profileImageFilePath: undefined };

  const notifications = isLogin ? await getNotificationsTop30(cookie) : [];

  return (
    <ScrollAnimationCover>
      <div className="flex justify-center relative items-center py-4 px-1 sm:px-5 max-w-screen-xl m-auto h-[70px]">
        <LogoButton />
        <Navigater currentPage={currentPage} />
        {isLogin ? (
          <Submenu1
            authority={authority}
            profileImageFilePath={profileImageFilePath}
            notifications={notifications}
          />
        ) : (
          <Submenu2 />
        )}
      </div>
      <NavigatorInMobile currentPage={currentPage} />
    </ScrollAnimationCover>
  );
}
