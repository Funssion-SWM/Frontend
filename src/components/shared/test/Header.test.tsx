import renderer from 'react-test-renderer';
import Header from '../Header';

jest.mock('next/navigation', () => jest.requireActual('next-router-mock'));

describe('Header', () => {
  it('renders correctly when isLogin is false', () => {
    const component = renderer.create(
      <Header isLogin={false} notifications={[]} profileImageFilePath={undefined} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders correctly when isLogin is true', () => {
    const component = renderer.create(
      <Header isLogin={true} notifications={[]} profileImageFilePath={undefined} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
