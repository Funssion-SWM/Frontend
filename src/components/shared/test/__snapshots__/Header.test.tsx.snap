// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly when isLogin is false 1`] = `
<section
  className="border-b-2 sticky top-0 bg-white z-10"
>
  <header
    className="flex justify-between items-center py-4 px-5 max-w-screen-xl m-auto"
  >
    <h1
      className="text-2xl font-bold cursor-pointer"
      onClick={[Function]}
    >
      Inforum
    </h1>
    <a
      href="/login"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      <button
        className="bg-soma-blue-40 text-white px-3.5 py-2 rounded-3xl transition hover:bg-soma-blue-50 text-xs sm:text-sm   "
        onClick={[Function]}
      >
        로그인
      </button>
    </a>
  </header>
</section>
`;

exports[`Header renders correctly when isLogin is true 1`] = `
<section
  className="border-b-2 sticky top-0 bg-white z-10"
>
  <header
    className="flex justify-between items-center py-4 px-5 max-w-screen-xl m-auto"
  >
    <h1
      className="text-2xl font-bold cursor-pointer"
      onClick={[Function]}
    >
      Inforum
    </h1>
    <nav
      className="flex items-center gap-3 relative"
    >
      <img
        alt="search_icon"
        className="cursor-pointer"
        data-nimg="1"
        decoding="async"
        height={40}
        loading="lazy"
        onClick={[Function]}
        onError={[Function]}
        onLoad={[Function]}
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcSet="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style={
          {
            "color": "transparent",
          }
        }
        width={40}
      />
      <button
        onClick={[Function]}
      >
        <img
          alt="profileImg"
          className="rounded-full w-8 h-8 object-cover"
          data-nimg="1"
          decoding="async"
          height={32}
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
          srcSet="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
          style={
            {
              "color": "transparent",
            }
          }
          width={32}
        />
      </button>
      <nav
        className="absolute top-10 bg-white flex flex-col gap-1 rounded-lg shadow-inner z-10 invisible"
      >
        <button
          className="hover:bg-gray-200 p-2 rounded-t-lg"
          onClick={[Function]}
        >
          프로필
        </button>
        <button
          className="hover:bg-gray-200 p-2 rounded-t-lg"
          onClick={[Function]}
        >
          임시 글
        </button>
        <button
          className="hover:bg-gray-200 p-2 rounded-b-lg"
          onClick={[Function]}
        >
          로그아웃
        </button>
      </nav>
      <button
        className="bg-soma-blue-40 text-white px-3.5 py-2 rounded-3xl transition hover:bg-soma-blue-50 text-xs sm:text-sm   "
        onClick={[Function]}
      >
        글쓰기
      </button>
    </nav>
  </header>
</section>
`;
